### PR TITLE
37 Setup dataloaders to sample new simulations

### PR DESCRIPTION
**Pull request by DeaglanBartlett**
Add dataloader which simulates instead of reads from file. Changes:

- New `SBISimulator` class in [loaders.py](https://github.com/maho3/ltu-ili/blob/main/ili/dataloaders/loaders.py) which can run simulations
- Added `num_rounds` and `x_obs` arguments to `__call__` function of [`SBIRunner`](https://github.com/maho3/ltu-ili/blob/main/ili/inference/runner_sbi.py#L22) to allow for multi-round inference
- Changed default weighting of ensemble to ensure that weights are always positive (problematic for active sampling if not)
- Observed `x` now given as argument to metrics so that sampling occurs with knowledge of this point
- Added new example to show how to use sbi's active sampler for toy example

### CODE REVIEW

File: ili/validation/runner.py [hunk 1]
```diff
@@ -72,19 +74,32 @@ def load_posterior(cls, path):
...
+            x_obs (torch.Tensor): tensor of the observed x
+            theta_obs (torch.Tensor): tensor of the true theta
         """
         t0 = time.time()
 
         # NOTE: sbi posteriors only accept torch.Tensor inputs
-        x_test = torch.Tensor(loader.get_all_data())
-        theta_test = torch.Tensor(loader.get_all_parameters())
+        if hasattr(loader, 'simulate'):
+            theta_test, x_test = loader.simulate(prior)
```
Comment (maho3) [inline ili/validation/runner.py:97]: If prior is None, raise Exception

**Review by maho3 — CHANGES_REQUESTED** [review]
Some small notes:
- I went ahead and pushed a removal of examples/ltu-ili-data since it was a symbolic link to a directory on infinity
- Should configs/val/sample.yaml and configs/val/multiround.yaml be merged since they're the same file?
- runner_sbi.py is getting hard to understand. Could you add quelque inline comments in __call__()?
- I left inline comments throughout Files Changed with minor fixes.

Bigger notes:
- Should x_obs and theta_obs be arguments that need be specified in the script example_*.py? I'm of the opinion that they shouldn't, as e.g. when you're doing inference on obs, you will want a loader to load them from file and not to have to build a loader -> loader.get_all_data -> submit them as parameters to runner and val_runner. Perhaps a more elegant solution is to have train and test loaders as arguments in runner and val_runner. What do you think?
- Before we close this, we should produce a comparison plot of posteriors trained with and without adaptive sampling, for a fixed number of simulations. The current toy example may be too simple to show much difference

File: ili/inference/runner_sbi.py [hunk 2]
```diff
@@ -115,38 +115,71 @@ def load_neural_posteriors(
...
+                    self.embedding_net.initalize_model(n_input=x.shape[-1])
+                density_estimator = all_model[n].append_simulations(theta, x, proposal).train(
+                        **self.train_args,
+                )
+                posteriors.append(all_model[n].build_posterior(density_estimator))
+                val_loss.append(all_model[n].summary["best_validation_log_prob"][-1])
+            val_loss = torch.tensor([float(vl) for vl in val_loss])
+            val_loss = torch.exp(val_loss - val_loss.max())
+            val_loss /= val_loss.sum()
+            print('\nVAL LOSS:', val_loss, '\n')
```
[the remark this answers is not in the document]
Comment (DeaglanBartlett) [reply ili/inference/runner_sbi.py:172]: No it's not. It will be removed in the next push

File: ili/inference/runner_sbi.py [hunk 3]
```diff
@@ -120,33 +125,59 @@ def __call__(self, loader):
...
+            if rnd > 0:
+                theta, x = loader.simulate(proposal)
+            posteriors, val_loss = [], []
+            for n, posterior in enumerate(self.neural_posteriors):
+                logging.info(
+                    f"Training model {n+1} out of {len(self.neural_posteriors)} ensemble models"
+                )
+                if not isinstance(self.embedding_net, nn.Identity):
+                    self.embedding_net.initalize_model(n_input=x.shape[-1])
+                density_estimator = all_model[n].append_simulations(theta, x, proposal).train(
```
Comment (maho3) [inline ili/inference/runner_sbi.py:164]: I checked and this does, by default, indeed always train the existing internal neural network every time we call train, without reinitialization ([link](https://github.com/mackelab/sbi/blob/5768a7c05aec799fe435167f75acaca483e5404d/sbi/inference/snpe/snpe_base.py#L304)). It also uses all training data that has been appended, so we only need pass the round-by-round theta's and x's. 

If you wanted a fresh neural network, you'd pass retrain_from_scratch=True. If you wanted to only use this round's data, you'd pass discard_prior_samples=True. Both of these are default = False.

Nothing to be done here, I just wanted to record this for future reference. :)

File: ili/dataloaders/loaders.py [hunk 4]
```diff
@@ -178,4 +180,94 @@ def load_parameters(
...
+        from sampling the proposal and simulating.
+
+        Args:
+            proposal (Any): Distribution to sample paramaters from
+
+        Returns:
+            Tuple[Tensor, Tensor]: Sampled parameters $\theta$ and simulation-outputs $x$.
+        """
+        theta, x = simulate_for_sbi(self.simulator, proposal, num_simulations=self.num_simulations)
+        np.save(self.theta_path, theta.detach().cpu().numpy())
```
[the remark this answers is not in the document]
Comment (maho3) [reply ili/dataloaders/loaders.py:249]: However, it should continue to only output the current round's data, as that is what `append_simulations` asks for.

File: ili/inference/runner_sbi.py
[the remark this answers is not in the document]
Comment (DeaglanBartlett) [reply ili/inference/runner_sbi.py:1]: I agree that is better. Will be added in next commit

File: ili/inference/runner_sbi.py [hunk 5]
```diff
@@ -120,33 +125,59 @@ def __call__(self, loader):
...
+        else:
+            x = torch.Tensor(loader.get_all_data())
+            theta = torch.Tensor(loader.get_all_parameters())
+            if self.num_rounds != 1:
+                raise Exception("If not simulating data you cannot have multi-round inference")
+        all_model = []
         for n, posterior in enumerate(self.neural_posteriors):
+            all_model.append(self.inference_class(
+                    prior=self.prior,
+                    device=self.device,
```
[the remark this answers is not in the document]
Comment (DeaglanBartlett) [reply ili/inference/runner_sbi.py:183]: I'm not sure why I removed this before. I have re-added it in the latest commit

**Review by maho3 — APPROVED** [review]

**Pull request merged**